proposals/42-human-readable-mutable-names.md

```diff
@@ -0,0 +1,235 @@
+    - User agency around DNS is respected: every code path doing a DNS lookup (resolving DNSLink, `/dnsaddr`, `/dns*` multiaddrs) should use either explicit or implicit DNS resolvers defined in IPFS node configuration file.
+      - By default OS-level DNS resolver is used as catch-all, but the user can delegate resolution of all or specific TLDs to specific DoH resolvers.
+      - Node tries to resolve with the most specific resolver (if present for TLD), and then fall back to the global one.
+    - Visual Aid / config mock-up
+      ```json
+        "DNS": {
+          "Resolvers": {
+            "*.eth": "https://eth.link-or-cloudflare/path/to/dns-query",
+            "*.crypto": "https://unstoppablesomething.example.com/path/to/dns-query",
+            "*.libre": "https://www.opennic.org/path/to/their/dns-query", 
+            "*": "https://mozilla.cloudflare-dns.com/dns-query"
+          }
+        }
+      ```
+- DNSLink provides interop for human-readable names and enables independence from PKI and ICANN.
+    - Anyone can set up and run own naming system without our involvement, blessing or endorsement.
+    - No vendor-specific client is included. 
+- IPNS provides a reliable solution for cryptographically-verifiable mutable names that won't break, as long someone is re-publishing the latest record.
+    - IPNS re-publishing works seamlessly, out of the box, just like providing CIDs for blocks in local datastore (pinned or not).
+    - It is possible to pin IPNS address to keep it around and ensure it resolves even when original publisher is gone.
+        - Node follows IPNS record updates and keeps re-publishing latest one if original author disappears.
+        - Node follows content updates behind the IPNS record, and update content pin every time IPNS record changes. 
+            - It is possible to opt-out from this behavior to re-publish IPNS record for `en.wikipedia-on-ipfs.org` without pinning all the data.
+
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+<!--
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+
+- We see % of `dnslink=/ipns/` going up.
+- We see more IPNS records on DHT
+- We no longer see requests for supporting new naming systems.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+ 
+- Configurable per TLDs [DNS over HTTPS](https://en.wikipedia.org/wiki/DNS_over_HTTPS) resolver for  may be not enough to mitigate concerns around centralized nature of ICANN's DNS.
+- IPNS may have other problems that block it from practical use in DNSLink records (eg. resolution still taking way longer than immutable path, making initial page load slow).
```

Comment:
Do we have any indicating data on what the performance will be like when https://github.com/protocol/web3-dev-team/pull/19 gets done (e.g., IPNS has PubSub)?  I assume @aschmahmann may know?

Reply:
iirc ipns-over-pubsub is no longer slowed down by DHT on the initial lookup (as long both resolver and publisher have it enabled), so it should be comparable with immutable paths (even if a bit slower, user won’t notice it because DHT  for resolving final path will take way more)